File: src/main/java/maifetch/json/Json.java
package maifetch.json;

import java.io.IOException;
import java.io.Reader;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class Json {
    private Json() {
    }

    public static Object parse(String input) {
        Parser parser = new Parser(input == null ? "" : input);
        Object value = parser.parseValue();
        parser.skipWhitespace();
        if (!parser.atEnd()) {
            throw new IllegalArgumentException("unexpected trailing JSON content");
        }
        return value;
    }

    public static Object parse(Reader reader) throws IOException {
        StringBuilder builder = new StringBuilder();
        char[] buffer = new char[4096];
        int read;
        while ((read = reader.read(buffer)) != -1) {
            builder.append(buffer, 0, read);
        }
        return parse(builder.toString());
    }

    private static final class Parser {
        private final String input;
        private int index;

        private Parser(String input) {
            this.input = input;
        }

        private Object parseValue() {
            skipWhitespace();
            if (atEnd()) {
                throw new IllegalArgumentException("unexpected end of JSON");
            }
            char ch = input.charAt(index);
            if (ch == '{') {
                return parseObject();
            }
            if (ch == '[') {
                return parseArray();
            }
            if (ch == '"') {
                return parseString();
            }
            if (ch == 't') {
                expect("true");
                return Boolean.TRUE;
            }
            if (ch == 'f') {
                expect("false");
                return Boolean.FALSE;
            }
            if (ch == 'n') {
                expect("null");
                return null;
            }
            if (ch == '-' || Character.isDigit(ch)) {
                return parseNumber();
            }
            throw new IllegalArgumentException("unexpected JSON character: " + ch);
        }

        private Map<String, Object> parseObject() {
            expect('{');
            Map<String, Object> object = new LinkedHashMap<String, Object>();
            skipWhitespace();
            if (peek('}')) {
                expect('}');
                return object;
            }
            while (true) {
                skipWhitespace();
                if (!peek('"')) {
                    throw new IllegalArgumentException("object key must be a string");
                }
                String key = parseString();
                skipWhitespace();
                expect(':');
                object.put(key, parseValue());
                skipWhitespace();
                if (peek('}')) {
                    expect('}');
                    return object;
                }
                expect(',');
            }
        }

        private List<Object> parseArray() {
            expect('[');
            List<Object> values = new ArrayList<Object>();
            skipWhitespace();
            if (peek(']')) {
                expect(']');
                return values;
            }
            while (true) {
                values.add(parseValue());
                skipWhitespace();
                if (peek(']')) {
                    expect(']');
                    return values;
                }
                expect(',');
            }
        }

        private String parseString() {
            expect('"');
            StringBuilder builder = new StringBuilder();
            while (!atEnd()) {
                char ch = input.charAt(index++);
                if (ch == '"') {
                    return builder.toString();
                }
                if (ch != '\\') {
                    builder.append(ch);
                    continue;
                }
                if (atEnd()) {
                    throw new IllegalArgumentException("unterminated JSON escape");
                }
                char escaped = input.charAt(index++);
                if (escaped == '"' || escaped == '\\' || escaped == '/') {
                    builder.append(escaped);
                } else if (escaped == 'b') {
                    builder.append('\b');
                } else if (escaped == 'f') {
                    builder.append('\f');
                } else if (escaped == 'n') {
                    builder.append('\n');
                } else if (escaped == 'r') {
                    builder.append('\r');
                } else if (escaped == 't') {
                    builder.append('\t');
                } else if (escaped == 'u') {
                    builder.append(parseUnicodeEscape());
                } else {
                    throw new IllegalArgumentException("unsupported JSON escape: " + escaped);
                }
            }
            throw new IllegalArgumentException("unterminated JSON string");
        }

        private char parseUnicodeEscape() {
            if (index + 4 > input.length()) {
                throw new IllegalArgumentException("incomplete unicode escape");
            }
            String hex = input.substring(index, index + 4);
            index += 4;
            try {
                return (char) Integer.parseInt(hex, 16);
            } catch (NumberFormatException error) {
                throw new IllegalArgumentException("invalid unicode escape: " + hex, error);
            }
        }

        private Number parseNumber() {
            int start = index;
            if (peek('-')) {
                index++;
            }
            readDigits();
            if (peek('.')) {
                index++;
                readDigits();
            }
            if (peek('e') || peek('E')) {
                index++;
                if (peek('+') || peek('-')) {
                    index++;
                }
                readDigits();
            }
            String raw = input.substring(start, index);
            if (raw.indexOf('.') >= 0 || raw.indexOf('e') >= 0 || raw.indexOf('E') >= 0) {
                return new BigDecimal(raw);
            }
            try {
                return Integer.valueOf(raw);
            } catch (NumberFormatException ignored) {
                return Long.valueOf(raw);
            }
        }

        private void readDigits() {
            int start = index;
            while (!atEnd() && Character.isDigit(input.charAt(index))) {
                index++;
            }
            if (start == index) {
                throw new IllegalArgumentException("expected JSON number digit");
            }
        }

        private void expect(String text) {
            if (!input.startsWith(text, index)) {
                throw new IllegalArgumentException("expected " + text);
            }
            index += text.length();
        }

        private void expect(char ch) {
            if (atEnd() || input.charAt(index) != ch) {
                throw new IllegalArgumentException("expected JSON character: " + ch);
            }
            index++;
        }

        private boolean peek(char ch) {
            return !atEnd() && input.charAt(index) == ch;
        }

        private void skipWhitespace() {
            while (!atEnd()) {
                char ch = input.charAt(index);
                if (ch != ' ' && ch != '\n' && ch != '\r' && ch != '\t') {
                    return;
                }
                index++;
            }
        }

        private boolean atEnd() {
            return index >= input.length();
        }
    }
}

File: src/main/kotlin/maifetch/json/Json.kt
package maifetch.json

import java.io.Reader
import java.math.BigDecimal

object Json {
    fun parse(input: String?): Any? {
        val parser = Parser(input.orEmpty())
        val value = parser.parseValue()
        parser.skipWhitespace()
        if (!parser.atEnd()) {
            throw IllegalArgumentException("unexpected trailing JSON content")
        }
        return value
    }

    fun parse(reader: Reader): Any? = parse(reader.readText())

    private class Parser(private val input: String) {
        private var index = 0

        fun parseValue(): Any? {
            skipWhitespace()
            if (atEnd()) throw IllegalArgumentException("unexpected end of JSON")
            return when (val ch = input[index]) {
                '{' -> parseObject()
                '[' -> parseArray()
                '"' -> parseString()
                't' -> {
                    expect("true")
                    true
                }
                'f' -> {
                    expect("false")
                    false
                }
                'n' -> {
                    expect("null")
                    null
                }
                '-', in '0'..'9' -> parseNumber()
                else -> throw IllegalArgumentException("unexpected JSON character: $ch")
            }
        }

        private fun parseObject(): Map<String, Any?> {
            expect('{')
            val result = linkedMapOf<String, Any?>()
            skipWhitespace()
            if (peek('}')) {
                expect('}')
                return result
            }
            while (true) {
                skipWhitespace()
                if (!peek('"')) throw IllegalArgumentException("object key must be a string")
                val key = parseString()
                skipWhitespace()
                expect(':')
                result[key] = parseValue()
                skipWhitespace()
                if (peek('}')) {
                    expect('}')
                    return result
                }
                expect(',')
            }
        }

        private fun parseArray(): List<Any?> {
            expect('[')
            val values = mutableListOf<Any?>()
            skipWhitespace()
            if (peek(']')) {
                expect(']')
                return values
            }
            while (true) {
                values += parseValue()
                skipWhitespace()
                if (peek(']')) {
                    expect(']')
                    return values
                }
                expect(',')
            }
        }

        private fun parseString(): String {
            expect('"')
            return buildString {
                while (!atEnd()) {
                    val ch = input[index++]
                    if (ch == '"') return@buildString
                    if (ch != '\\') {
                        append(ch)
                        continue
                    }
                    if (atEnd()) throw IllegalArgumentException("unterminated JSON escape")
                    when (val escaped = input[index++]) {
                        '"', '\\', '/' -> append(escaped)
                        'b' -> append('\b')
                        'f' -> append('\u000C')
                        'n' -> append('\n')
                        'r' -> append('\r')
                        't' -> append('\t')
                        'u' -> append(parseUnicodeEscape())
                        else -> throw IllegalArgumentException("unsupported JSON escape: $escaped")
                    }
                }
                throw IllegalArgumentException("unterminated JSON string")
            }
        }

        private fun parseUnicodeEscape(): Char {
            if (index + 4 > input.length) {
                throw IllegalArgumentException("incomplete unicode escape")
            }
            val hex = input.substring(index, index + 4)
            index += 4
            return hex.toInt(16).toChar()
        }

        private fun parseNumber(): Number {
            val start = index
            if (peek('-')) index++
            readDigits()
            if (peek('.')) {
                index++
                readDigits()
            }
            if (peek('e') || peek('E')) {
                index++
                if (peek('+') || peek('-')) index++
                readDigits()
            }
            val raw = input.substring(start, index)
            if ('.' in raw || 'e' in raw || 'E' in raw) return BigDecimal(raw)
            return raw.toIntOrNull() ?: raw.toLong()
        }

        private fun readDigits() {
            val start = index
            while (!atEnd() && input[index].isDigit()) {
                index++
            }
            if (start == index) {
                throw IllegalArgumentException("expected JSON number digit")
            }
        }

        private fun expect(text: String) {
            if (!input.startsWith(text, index)) throw IllegalArgumentException("expected $text")
            index += text.length
        }

        private fun expect(ch: Char) {
            if (atEnd() || input[index] != ch) throw IllegalArgumentException("expected JSON character: $ch")
            index++
        }

        private fun peek(ch: Char): Boolean = !atEnd() && input[index] == ch

        fun skipWhitespace() {
            while (!atEnd() && input[index] in charArrayOf(' ', '\n', '\r', '\t')) {
                index++
            }
        }

        fun atEnd(): Boolean = index >= input.length
    }
}
